Improved patterns command with new algorithm

songkant-aws opened the pull request:
### Description
This PR introduces enhancement on original patterns command. It keeps the patterns command naming and rebuild it with specific window functions to parse log messages with different pattern method(log parser algorithms). See this RFC: https://github.com/opensearch-project/sql/issues/3251

The sample query input and output will be like:

* New brain algorithm log parser:
<img width="1458" alt="Screenshot 2025-02-14 at 14 22 01" src="https://github.com/user-attachments/assets/4bd155ff-e6cc-4ceb-a9b0-670be318a450" />

* Original regex based patterns command:
<img width="1444" alt="Screenshot 2025-02-14 at 14 22 21" src="https://github.com/user-attachments/assets/c4a7151a-b2b5-45b8-a17e-119259ec7ac9" />

* After changing default pattern method to brain, the default behavior is changed:
<img width="1454" alt="Screenshot 2025-02-14 at 14 23 15" src="https://github.com/user-attachments/assets/24c7aaa8-5f5a-44ee-84bf-16791c5ab22c" />

### Related Issues
Resolves #3251 

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
 - [ ] New functionality has javadoc added.
 - [x] New functionality has a user manual doc added.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

YANG-DB commented:
Thanks for the initiative!!

penghuo commented on ppl/src/main/java/org/opensearch/sql/ppl/parser/AstBuilder.java:
```diff
@@ -283,18 +289,35 @@ public UnresolvedPlan visitParseCommand(OpenSearchPPLParser.ParseCommandContext
+                new Function(
+                    ctx.pattern_method != null
+                        ? ctx.pattern_method.getText().toLowerCase(Locale.ROOT)
+                        : BuiltinFunctionName.BRAIN.name(), // By default, use new algorithm
```
Could we introduce a setting to control the default pattern algorithms?

songkant-aws replied:
Ok, I see. Like a java properties to specify default algorithm name? I can add that.

penghuo commented on core/src/main/java/org/opensearch/sql/expression/window/patterns/StreamPatternWindowFunction.java:
```diff
@@ -0,0 +1,80 @@
+                    ((NamedArgumentExpression) expression).getValue().valueOf().stringValue())
+            .findFirst()
+            .orElse("");
+    return new StreamPatternRowWindowFrame(
```
do we need StreamPatternRowWindowFrame? could we just use CurrentRowWindowFrame instead? patternExpression could be member of StreamPatternWindowFunction?

songkant-aws replied:
Sure, I will remove the need of StreamPatternRowWindowFrame in the next revision.

penghuo commented on core/src/main/java/org/opensearch/sql/expression/window/frame/PeerRowsWindowFrame.java:
```diff
@@ -109,20 +118,15 @@ public void load(PeekingIterator<ExprValue> it) {
   }
 
-  private boolean isSamePartition(ExprValue next) {
+  protected boolean isSamePartition(ExprValue next) {
```
revert unnecessary change.

songkant-aws replied:
Done

penghuo commented on core/src/main/java/org/opensearch/sql/expression/window/frame/BufferPatternRowsWindowFrame.java:
```diff
@@ -0,0 +1,62 @@
+                  return value.stringValue();
+                })
+            .toList();
+    this.preprocessedMessages.addAll(logParser.preprocessAllLogs(logMessages));
```
BufferPatternRowsWindowFrame should have it own spec, doet it means over all rows? @dai-chen  WindowFrame definition and How to use WIndowFrame should be seperate, right?

songkant-aws replied:
For simplicity, I set the spec to empty `partition by` and empty `sort by` during AST tree parsing unresolved WindowFunction, which treats the window frame range is all rows on coordinator node. Because I haven't seen requirements on sorting and partitioning on other columns. 

I think we can add `partition by` and `sort by` syntax if we see values in case users want to specify them. Thoughts?

LantaoJin replied:
Tests with non-empty `partition by` and non-empty `sort by` are still required in PR.

songkant-aws replied:
Added non-empty `partition by` and `sort by` unit tests.

penghuo reviewed (commented):
In general, 
* add user facing doc to explain new BRAIN alg with example will be helpfull.
* add UT could help reviwer understand the logic also. @dai-chen did a good example. https://github.com/opensearch-project/sql/blob/main/core/src/main/java/org/opensearch/sql/expression/window/frame/PeerRowsWindowFrame.java
* add benchmark to compare BRAIN and SIMPLE also help a lot.

dai-chen commented on core/src/main/java/org/opensearch/sql/expression/window/WindowFunctions.java:
```diff
@@ -34,6 +38,8 @@ public void register(BuiltinFunctionRepository repository) {
     repository.register(rowNumber());
     repository.register(rank());
     repository.register(denseRank());
+    repository.register(brain());
+    repository.register(simplePattern());
```
I'm thinking are these new algorithm real window function? If users specify `order by` or `partition by`, does it still generate meaningful result?

dai-chen commented on core/src/main/java/org/opensearch/sql/ast/dsl/AstDSL.java:
```diff
@@ -489,6 +491,22 @@ public static Parse parse(
+      UnresolvedExpression sourceField,
+      String alias,
+      java.util.Map<String, Literal> arguments) {
+    return new Pattern(
```
Just wonder if pattern is mostly syntax sugar for pattern window function, is new logical operator still required?

songkant-aws replied:
After the integration with LogicalWindow, I see they are quite similar. Yeah, I think Pattern operator is probably not needed.

songkant-aws replied:
I don't find the Window unresolvedPlan. So replace Pattern with Window instead.

dai-chen commented on core/src/main/java/org/opensearch/sql/expression/window/patterns/BufferPatternWindowFunction.java:
```diff
@@ -0,0 +1,88 @@
+import org.opensearch.sql.expression.window.frame.WindowFrame;
+
+@EqualsAndHashCode(callSuper = true)
+public class BufferPatternWindowFunction extends FunctionExpression
```
`order by` expression in window definition decides who're peers right? What's the order by expression for buffer offline pattern function? I didn't find it and may miss it in design doc:

```
Project[message#1, patterns_field#2]
+- Window[brain(message#1), windowsSpec(partitionBy=null, frame=PeerRowsWindowFrame)]
   +- OpenSearchIndexScan
```

dai-chen reviewed (commented):
A high-level question about the long-term approach: Currently, the pattern command always generates a new field, which offers flexibility as users can perform aggregation (stats) or other operations on it afterward. However, for large datasets, I assume most users would prefer functionality similar to [CloudWatch Logs Query Syntax - Pattern](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL_QuerySyntax-Pattern.html).

Do you see any potential issues with the current implementation, which is based on SQL window functions combined with aggregate functions, as we scale this feature in the future?

songkant-aws commented:
> Do you see any potential issues with the current implementation, which is based on SQL window functions combined with aggregate functions, as we scale this feature in the future?

@dai-chen It's a good question. Actually, it's one of my previous thought to use a new specific operator to directly return grouped logs with sample count or grouped samples.

It's more like a compound aggregation operator with a default `group by`. But current aggregation abstraction only supports row by row iteration. The log pattern algorithms cover both streaming and buffering computing paradigm. Then I figure out this is more close to window operator's abstraction. To achieve what Cloudwatch has done, I think we could extend pattern functions from `AggregateWindowFunction` that has a special aggregator state structure.

As to `partition by` and `sort by` specification, my initial plan is to deliver a simple version without them because that's probably enough.
`partition by` is probably more useful than `sort by`. Imagine users may want to see different log patterns for INFO, WARN, ERROR partitions.
I haven't seen a use case for `sort by` yet. Maybe they need to see latest logs `sort by` timestamp?

songkant-aws commented:
To scale the pattern function, I think other issues are pending investigation.

1. How to push down operator to OpenSearch data node? Is Painless script one of option?
2. OpenSearch doesn't have shuffle mechanism for now, how to perform large dataset aggregation like operation with simple map reduce? If data volume is quite large, spilling additional data to disk is probably required during execution.

songkant-aws commented:
AvgTime benchmark:

```
Benchmark                                          (testDataType)  Mode  Cnt  Score    Error  Units
ComparisonOperatorBenchmark.testEqualOperator                 int  avgt    3  0.002 ±  0.001  ms/op
ComparisonOperatorBenchmark.testEqualOperator              string  avgt    3  0.005 ±  0.001  ms/op
ComparisonOperatorBenchmark.testEqualOperator                date  avgt    3  0.001 ±  0.001  ms/op
ComparisonOperatorBenchmark.testGreaterOperator               int  avgt    3  0.002 ±  0.001  ms/op
ComparisonOperatorBenchmark.testGreaterOperator            string  avgt    3  0.005 ±  0.002  ms/op
ComparisonOperatorBenchmark.testGreaterOperator              date  avgt    3  0.001 ±  0.001  ms/op
ComparisonOperatorBenchmark.testLessOperator                  int  avgt    3  0.002 ±  0.001  ms/op
ComparisonOperatorBenchmark.testLessOperator               string  avgt    3  0.006 ±  0.001  ms/op
ComparisonOperatorBenchmark.testLessOperator                 date  avgt    3  0.002 ±  0.001  ms/op
PatternsWindowFunctionBenchmark.testBrain                     N/A  avgt    3  0.131 ±  0.002  ms/op
PatternsWindowFunctionBenchmark.testSimplePattern             N/A  avgt    3  0.001 ±  0.001  ms/op
```

songkant-aws commented:
This PR still needs some action items.
1. Pending adding setting to switch default log pattern algorithm
2. Pending discussion whether we want to deliver `partition by` and `order by` given its effect. (Log patterns are probably special window function that is not sensitive to `partition by` and `order by`)

LantaoJin commented on docs/user/ppl/cmd/patterns.rst:
```diff
@@ -12,24 +12,45 @@ patterns
+
+Simple Pattern
+============
+patterns [new_field=<new-field-name>] [pattern=<pattern>] <field> SIMPLE_PATTERN
```
The seems a breaking change, why not make `SIMPLE_PATTERN` optional for compatibility.

songkant-aws replied:
It's still compatible. If we omit SIMPLE_PATTERN, it will use BRAIN method by default. Peng suggested we can add a configuration to decide which is the default behavior.

LantaoJin replied:
> It's still compatible. If we omit `SIMPLE_PATTERN`, it will use BRAIN method by default. Peng suggested we can add a configuration to decide which is the default behavior.

We don't call it compatible if it brings breaking change with original command syntax. I doubt about using the new BRAIN algorithm as default output before it has been running stably for a while. @penghuo any thoughts here?

penghuo replied:
> Peng suggested we can add a configuration to decide which is the default behavior.
Did we add setting already? 
If not, we should default to SIMPLE_PATTERN.

songkant-aws replied:
Add a new setting now. It will default to SIMPLE_PATTERN. Please review the change. Let me know if it makes sense.

LantaoJin commented on ppl/src/main/antlr/OpenSearchPPLParser.g4:
```diff
@@ -115,18 +115,25 @@ parseCommand
-patternsMethod
-   : PUNCT
-   | REGEX
+patternMethod
+   : SIMPLE_PATTERN
+   | BRAIN
```
`patternMethod` should be added to `keywordsCanBeId` as well.

songkant-aws replied:
Done

LantaoJin commented on ppl/src/main/antlr/OpenSearchPPLParser.g4:
```diff
@@ -115,18 +115,25 @@ parseCommand
    : PARSE (source_field = expression) (pattern = stringLiteral)
    ;
 
+patternsMethod
+   : PUNCT
+   | REGEX
```
`patternsMethod` should be added to `keywordsCanBeId`

songkant-aws replied:
Done

penghuo commented:
@songkant-aws could u merge conflict?

songkant-aws commented:
@penghuo @dai-chen @LantaoJin I have addressed several comments and added a new setting `plugins.ppl.default.pattern.method` to allow overriding default pattern method. For now, the default pattern method is still regex based SIMPLE_PATTERN. Please let me know if the current change is reasonable. Thanks!

penghuo commented:
@songkant-aws please fix failed UT coverage. then we are ready to merge.

songkant-aws commented:
@penghuo Fixed all of CI checks with more test coverage and cases.

penghuo reviewed (approved):
Thx!

LantaoJin commented on ppl/src/main/java/org/opensearch/sql/ppl/parser/AstBuilder.java:
```diff
@@ -283,18 +292,35 @@ public UnresolvedPlan visitParseCommand(OpenSearchPPLParser.ParseCommandContext
-    ImmutableMap.Builder<String, Literal> builder = ImmutableMap.builder();
+    List<UnresolvedExpression> unresolvedArguments = new ArrayList<>();
+    unresolvedArguments.add(sourceField);
+    AtomicReference<String> alias = new AtomicReference<>("patterns_field");
```
`AstBuilder` is created per query. No necessary to wrap with `AtomicReference`

songkant-aws replied:
Seems like IDE has compilation error when using a temp String variable in lambda expression. So I use AtomicReference as a workaround

LantaoJin replied:
Ok, I see, the `String alias` must be a "final" variable from the enclosing scope.

LantaoJin commented on common/src/main/java/org/opensearch/sql/common/patterns/BrainLogParser.java:
```diff
@@ -0,0 +1,443 @@
+      float thresholdPercentage,
+      Map<Pattern, String> filterPatternVariableMap,
+      List<String> delimiters) {
+    if (thresholdPercentage < 0.0f || thresholdPercentage > 1.0f) {
+      throw new IllegalArgumentException("Threshold percentage must be between 0.0 and 1.0");
+    }
```
What if the `thresholdPercentage = 0` or `thresholdPercentage = 1`

songkant-aws replied:
It's valid, which means 0 or the highest frequency.

LantaoJin commented on core/src/main/java/org/opensearch/sql/expression/window/patterns/BufferPatternWindowFunction.java:
```diff
@@ -0,0 +1,70 @@
+import org.opensearch.sql.expression.window.frame.WindowFrame;
+import org.opensearch.sql.utils.FunctionUtils;
+
+@EqualsAndHashCode(callSuper = true)
```
minor: `@ToString`

songkant-aws replied:
I override this function's toString specifically.

LantaoJin commented on core/src/main/java/org/opensearch/sql/expression/window/patterns/StreamPatternWindowFunction.java:
```diff
@@ -0,0 +1,70 @@
+import org.opensearch.sql.expression.window.frame.WindowFrame;
+import org.opensearch.sql.utils.FunctionUtils;
+
+@EqualsAndHashCode(callSuper = true)
```
ditto
